rcon/player_history.py

```diff
@@ -275,10 +294,31 @@ def add_player_to_blacklist(steam_id_64, reason, name=None):
                     steamid=player, is_blacklisted=True, reason=reason)
             )
 
+        try:
+            if rcon and name and ban_if_in_game:
+                if name in [p['name'] for p in rcon.get_players()]:
+                    info = rcon.get_player_info(name)
+                    if info['steam_id_64'] == steam_id_64:
+                        ban_and_add_log(rcon, info['steam_id_64'], name, reason)
+        except:
+                logger.warning("Something went wrong banning player with steam id %s", steam_id_64)
+
         sess.commit()
 
 
-def remove_player_from_blacklist(steam_id_64):
+def _find_and_remove_perma_bans_by_ban_log(rcon, ban_logs, sess):
+    counter = 0
+    perma_bans = rcon.get_perma_bans()
+    for ban_log_entry in ban_logs:
+        if ban_log_entry.ban_log:
+            if ban_log_entry.ban_log in perma_bans:
+                rcon.do_remove_perma_ban(ban_log_entry.ban_log, 'BLACKLIST_REMOVE', False)
```

Comment:
You'll want to invalidate the cache of perma bans for each call to a do_perma or do_remove_perma